pkg/etcdcli/health.go

```diff
@@ -43,34 +43,21 @@ type memberHealth []healthCheck
 
 func GetMemberHealth(ctx context.Context, etcdMembers []*etcdserverpb.Member) memberHealth {
 	memberHealth := memberHealth{}
+	resChan := make(chan healthCheck, 1)
 	for _, member := range etcdMembers {
 		if !HasStarted(member) {
 			memberHealth = append(memberHealth, healthCheck{Member: member, Healthy: false})
 			continue
 		}
 
-		const defaultTimeout = 30 * time.Second
-		resChan := make(chan healthCheck, 1)
 		go func() {
-			ctxTimeout, cancel := context.WithTimeout(ctx, defaultTimeout)
-			defer cancel()
-
-			resChan <- checkSingleMemberHealth(ctxTimeout, member)
-			// closing here to avoid late replies to panic on resChan,
-			// the result will be considered a timeout anyway
-			close(resChan)
```

Comment:
we kinda still have to close this channel, don't we? 

fyi, that's a panic I've fixed recently:
https://issues.redhat.com//browse/OCPBUGS-27959

https://github.com/openshift/cluster-etcd-operator/pull/1190

Reply:
Looking at the pre-#1190 [code][1], the panics were from timelines like:

1. `checkSingleMemberHealth` goroutine launched with its own 30s `Context` duration.
2. `select` waited on a result from `resChan` _or_ a new 30s `time.After`.
3. When the `select` `time.After` won, it appended a `30s timeout waiting for member...` to `memberHealth`, and `close`d `resChan`.
4. A millisecond or two later, `checkSingleMemberHealth` would hit its 30s `Context` timeout in [the `Get` call][2], create [its own `health check failed: ...` result][3], and [push it into `resChan`][4].
5. But `resChan` was closed in step 3!  Panic!

With #1190, you dropped the `close` from step 3, and moved it to step 4, so no more panic.

But from [Go's _Range and Close_ tour][5]:

> Channels aren't like files; you don't usually need to close them. Closing is only necessary when the receiver must be told there are no more values coming, such as to terminate a range loop.

And with _this_ pull, we no longer have the receiver-side `select` or timeout.  With this pull, the receiver will block until it has a result back from each launched `checkSingleMemberHealth` goroutine, and it's up to those goroutines to respect the `Context` timeout.  So there is no chance of the `GetMemberHealth` `close`-ing the channel before a `checkSingleMemberHealth` goroutine writes, because we no longer have an explicit `close` at all, and the channel is just garbage-collected as it goes out of scope, like all local Go variables.

[1]: https://github.com/openshift/cluster-etcd-operator/blob/1e1970f23b7a588953ad47617108a683ecc4f30f/pkg/etcdcli/health.go#L61-L72
[2]: https://github.com/openshift/cluster-etcd-operator/blob/1e1970f23b7a588953ad47617108a683ecc4f30f/pkg/etcdcli/health.go#L121-L124
[3]: https://github.com/openshift/cluster-etcd-operator/blob/1e1970f23b7a588953ad47617108a683ecc4f30f/pkg/etcdcli/health.go#L136C26-L136C45
[4]: https://github.com/openshift/cluster-etcd-operator/blob/1e1970f23b7a588953ad47617108a683ecc4f30f/pkg/etcdcli/health.go#L58
[5]: https://go.dev/tour/concurrency/4

Reply:
Thanks for the explanation, indeed kicking the select out and just looping all the values is enough :)